### PR TITLE
nixos/hardware/ipu6: update all related packages

### PR DESCRIPTION
**Pull request by Princemachiavelli**
ipu6-drivers: update to fix compilation with kernels >= 6.6.0.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by mweinelt** [comment]
Please also check for updates for other packages in that ecosystem:

- ipu6-camera-hal
- ipu6-camera-bins
- icamerasrc

**Comment by Princemachiavelli** [comment]
Thanks for identifying those related packages. I'm taking a look at the upstream changes to those packages. There's been a bit of a refactor to `ipu6-camera-bins` that I need to figure out and test on hardware later today.

**Comment by jonringer** [comment]
If you want to update this PR with the other updates, I can help test as well

**Comment by flokli** [comment]
There's been some updates here: https://github.com/NixOS/nixpkgs/issues/225743#issuecomment-1731593624

However, they seem to break other parts of the stack…

**Comment by flokli** [comment]
I tried applying just the patch from here, and while it builds, I can't get the camera to output any signal.

I didn't try cherry-picking in https://github.com/NixOS/nixpkgs/compare/master...andresilva:nixpkgs:ipu6-updates in addition to the changes from here.

**Comment by Princemachiavelli** [comment]
> I tried applying just the patch from here, and while it builds, I can't get the camera to output any signal.
> 
> I didn't try cherry-picking in [master...andresilva:nixpkgs:ipu6-updates](https://github.com/NixOS/nixpkgs/compare/master...andresilva:nixpkgs:ipu6-updates) in addition to the changes from here.

That's what I found as well. I had working video prior but also had an out of nixpkgs patch of `ivsc-firmware` which may have broke.

I have also updated the remaining Intel packages in my WIP branch (https://github.com/Princemachiavelli/nixpkgs/tree/linux_6_6_ipu6_wip) but I still get no video.

**Comment by flokli** [comment]
@Princemachiavelli would you mind cherry-picking all of this into this PR? It's probably easier to iterate on, might only require another bump of userspace stuff, and at least results in a building kernel module ;-)

**Comment by flokli** [comment]
Did you forget to do the package rename in the module system?

```

       error: undefined variable 'ipu6ep-camera-bin'

       at /nix/store/h9xyh49h3p7f6vicblc09pahpwrq721g-nixpkgs-src/nixos/modules/hardware/video/webcam/ipu6.nix:34:46:

           33|       ++ optional (cfg.platform == "ipu6") ipu6-camera-bin
           34|       ++ optional (cfg.platform == "ipu6ep") ipu6ep-camera-bin;
             |                                              ^
           35|
```

File: pkgs/os-specific/linux/firmware/ipu6-camera-bins/default.nix [hunk 1]
```diff
@@ -4,26 +4,19 @@
...
 , zlib
-
-# Pick one of
-# - ipu6 (Tiger Lake)
-# - ipu6ep (Alder Lake)
-, ipuVersion ? "ipu6"
 }:
 
 stdenv.mkDerivation (finalAttrs: {
-  pname = "${ipuVersion}-camera-bin";
```
Comment (andresilva) [inline pkgs/os-specific/linux/firmware/ipu6-camera-bins/default.nix:15]: IMO this should be renamed to `ipu6-camera-bins` (same as the upstream project) since it includes all the bins.

File: pkgs/os-specific/linux/firmware/ipu6-camera-bins/default.nix [hunk 2]
```diff
@@ -40,13 +33,13 @@ stdenv.mkDerivation (finalAttrs: {
...
-    install -m 0644 -D ../LICENSE $out/share/doc/LICENSE
+    install -m 0644 -D LICENSE $out/share/doc/LICENSE
 
     runHook postInstall
   '';
 
   postFixup = ''
-    for pcfile in $out/lib/pkgconfig/*.pc; do
+    for pcfile in $out/lib/*/pkgconfig/*.pc; do
       substituteInPlace $pcfile \
```
Comment (andresilva) [inline pkgs/os-specific/linux/firmware/ipu6-camera-bins/default.nix:43]: The only replace needed here is `prefix=/usr' "prefix=$out`, everything else is already correct upstream.

**Comment by andresilva** [comment]
In the last commit message you mention a Dell config to use as a guideline, can you link to what you're referring?

File: pkgs/os-specific/linux/firmware/ipu6-camera-bins/default.nix (hunk 1 above)
Comment (flokli) [reply pkgs/os-specific/linux/firmware/ipu6-camera-bins/default.nix:15]: I'd just rename it like it is called upstream and update the references TBH.

**Comment by andresilva** [comment]
Been doing some more testing. I think part of the issue is this:

```
❯ ldd /nix/store/j8miciqcamr4mk2f0q0vp5w0a33xz49s-icamerasrc-ipu6ep-unstable-2023-10-23/lib/gstreamer-1.0/libgsticamerasrc.so
...
	libia_cca.so => not found
	libia_aiq.so => not found
	libia_aiqb_parser.so => not found
	libia_cmc_parser.so => not found
	libia_exc.so => not found
	libia_mkn.so => not found
	libia_emd_decoder.so => not found
	libia_bcomp.so => not found
	libia_nvm.so => not found
	libia_log.so => not found
	libia_dvs.so => not found
	libia_coordinate.so => not found
	libia_ltm.so => not found
	libia_lard.so => not found
	libia_isp_bxt.so => not found
	libbroxton_ia_pal.so => not found
	libgcss.so.0 => not found
...
```
All of these libraries should come from `ipu6-camera-bins` and for some reason they're not being found while building icamerasrc.

**Comment by andresilva** [comment]
I can't get `v4l2-relayd` to work unfortunately. I updated `v4l2loopback` (although it wasn't needed), and can get the camera to work on firefox by writing into the loopback with gstreamer:

```
DEVICE=$(sudo v4l2loopback-ctl add -x 1 -n 'test')
sudo -E gst-launch-1.0 icamerasrc buffer-count=7 ! video/x-raw,width=1280,height=720,framerate=30/1 ! videoconvert ! v4l2sink device=$DEVICE
```

The camera immediately starts recording (the led turns on), and works on firefox.

**Comment by flokli** [comment]
> All of these libraries should come from `ipu6-camera-bins` and for some reason they're not being found while building icamerasrc (likely related with them being in a folder for the ipu target now).

Does it work if you patchelf them to these paths in the fixup phase, rather than trying to patch all consumers?

**Comment by andresilva** [comment]
I guess it would probably work but I didn't try it, I just wanted to move on and see the camera work :)

**Comment by andresilva** [comment]
@Princemachiavelli can I push some changes to this PR directly or should I push to another branch and let you cherry-pick?

**Comment by andresilva** [comment]
Here's a script to test the webcam: https://gist.github.com/andresilva/b019cedbc1f78f34f3537c9d95869639
The only assumption the script makes is that the ipu6 module is enabled.

**Comment by andresilva** [comment]
> Does it work if you patchelf them to these paths in the fixup phase, rather than trying to patch all consumers?

This worked, I have updated my branch to do this instead of setting `LD_LIBRARY_PATH`.

**Comment by flokli** [comment]
> > Does it work if you patchelf them to these paths in the fixup phase, rather than trying to patch all consumers?
> 
> This worked, I have updated my branch to do this instead of setting `LD_LIBRARY_PATH`.

Nice!  Does `v4l2-relayd` now work too, or is this still an issue?

**Comment by andresilva** [comment]
I actually hadn't tested, for some reason I assumed it wouldn't make any difference (compared to the `LD_LIBRARY_PATH` hack), but I was wrong! It's working fine now 🥳 On X1 Cargon Gen 11. I can either push all my changes to this PR or maybe create a new one.

**Comment by flokli** [comment]
I gave @andresilva's branch a try. Had to reboot so the v4lloopback kernel module got loaded again, but now it shows up and works in Google Meet (in Chromium) :tada: 

I noticed it does not work in the new GNOME Camera app ("Could not play camera stream"), but I'm not sure if it worked in previous versions either.

**Comment by andresilva** [comment]
I pushed all my changes now. There are some updates that aren't strictly necessary, namely `v4l2loopback`, `ivsc-driver` and `ivsc-firmware`. I can put those into a different PR if necessary but they're somewhat related to the ipu6 support.

File: nixos/modules/hardware/video/webcam/ipu6.nix [hunk 3]
```diff
@@ -44,14 +43,13 @@ in
...
       extraPackages = with pkgs.gst_all_1; [ ]
         ++ optional (cfg.platform == "ipu6") icamerasrc-ipu6
-        ++ optional (cfg.platform == "ipu6ep") icamerasrc-ipu6ep;
+        ++ optional (cfg.platform == "ipu6ep") icamerasrc-ipu6ep
+        ++ optional (cfg.platform == "ipu6epmtl") icamerasrc-ipu6epmtl;
 
       input = {
         pipeline = "icamerasrc";
-        format = mkIf (cfg.platform == "ipu6ep") (mkDefault "NV12");
+        format = mkIf (cfg.platform != "ipu6") (mkDefault "NV12");
```
Comment (flokli) [inline nixos/modules/hardware/video/webcam/ipu6.nix:51]: why this change?

**Comment by flokli** [comment]
@andresilva can you un-draft this?

File: nixos/modules/hardware/video/webcam/ipu6.nix (hunk 3 above)
Comment (andresilva) [reply nixos/modules/hardware/video/webcam/ipu6.nix:51]: We also support `ipu6epmtl` now, which should also use `NV12` AFAIU.

**Comment by andresilva** [comment]
Done. I also update the PR title but feel free to propose something else.

**Comment by andresilva** [comment]
Didn't expect this to cause so many rebuilds, we might need to target `staging`.

**Comment by flokli** [comment]
Hmmh, what's the culprit? v4l2loopback? If it still works with the older version, can we move this into a separate PR and keep the rest in this?

**Comment by andresilva** [comment]
The only thing depending on `v4l2loopback` besides `v4l2relayd` is `droidcam`. So I suspect the culprit is actually `icamerasrc` since it's part of gstreamer, it causes a rebuild of `gst_all_1` which a lot of things depend on. We can't update that on its own though :(

**Comment by flokli** [comment]
Alright, then let's retarget staging. Please draft again before doing the rebase, so we don't mass-ping the world.
